release-21.1: Backport changefeed observability PRs.

miretskiy opened the pull request:
Backport:
  * 4/4 commits from "changefeed,server: changefeed histograms and configurable slow logging" (#63923)
  * 2/2 commits from " changefeedccl: changefeed concurrency and frontier observability #67206" (#67268)

Please see individual PRs for details.

/cc @cockroachdb/release

Release Justification: A low danger, high impact observability change.  The added metrics allow us to troubleshoot
changefeed performance, when running in large scale clusters.

stevendanna committed:
changefeed: add new histogram metrics

This adds 3 new histogram metrics to try to get some insight into
whether we are seeing occasionally slow sinks or checkpoints during
various nightly roachtests.

While this produces some data duplication with the previous
flush_nanos, flushes, emitted_messages, and emit_nanos metrics, I
think that having the histogram will still be useful to more easily
see if a changefeed experienced a small number of slow flushes or
checkpoints.

The naming of these metrics is a bit unfortunate, but since
flush_nanos and emit_nanos already existed and since I didn't want to
remove them, I've included 'hist' in the name of these new metrics.

Release note: None

stevendanna committed:
changefeed: add slow_log_threshold cluster setting

This adds a new cluster setting

    changfeed.slow_span_log_threshold

That allows us to control the threshold for logging slow spans. This
is useful for cases where the auto-calculated threshold is much higher
than we would like.

Release note (sql change): A new cluster setting
`changefeed.slow_span_log_threshold` allows setting a cluster-wide
default for slow span logging.

stevendanna committed:
server: store total observations for histogram metrics

I occasionally find this useful to know how many observations a given
histogram is based on. The prometheus output already returns this, but
it is nice to have it in the SQL and JSON output as well.

Release note (ops change): Histogram metrics now store the total
number of observations over time.

Commit:
changfeedccl: Introduce a knob to control concurrency of scan requests.

Stop relying on ExportRequestLimit to determine the number of concurrent
export requests, and introduce a decidated ScanRequestLimit setting.

If the setting is specified, uses that setting; otherwise, the default
value is computed as 3 * (number of nodes in the cluster), which is the
old behavior, but we cap this number so that concurrency does not get
out of hand if running in a very large cluster.

Fixes #67190

Release Nodes: Provide a better configurability of scan request
concurrency.  Scan requests are issued by changefeeds during the
backfill.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68106)
<!-- Reviewable:end -->

stevendanna commented on pkg/settings/setting.go:
```diff
@@ -20,7 +20,7 @@ import (
 // MaxSettings is the maximum number of settings that the system supports.
 // Exported for tests.
-const MaxSettings = 512
+const MaxSettings = 288
```
Guessing we don't want to decrease this :D

miretskiy replied:
is this persisted somewhere?

stevendanna replied:
Not persisted, but if we register more settings than the max a panic is thrown.  My guess is a few PRs bumped it when we hit the old limit, so we should probably stick with the higher value.

miretskiy replied:
Fixed. Thanks.

miretskiy replied:
Removed one of the commits since the setting was increased multiple times.

Commit:
changefeedccl: Improve observability of change frontier updates.

Add a metric to keep track of the number of frontier updates in the
changefeed.

Release Notes: None

stevendanna reviewed (approved):
The metrics changes have been baking a while and are pretty low risk.  The changes to the export request limit seems pretty straightforward and high value for large clusters.
